core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java

```diff
@@ -61,6 +61,7 @@ private boolean refill() throws IOException {
         nRead = fileChannel.read(byteBuffer);
       }
       if (nRead < 0) {
+        byteBuffer.flip();
```

Comment:
This should be a bug.  The byteBuffer is fiped after `fileChannel.read(byteBuffer)`, so if then we can read the remaining  dirty data. The test case will be add latter. I'm not sure this is should be another patch. @cloud-fan

Reply:
Doesn't `nRead < 0` means we actually don't read any bytes?

Reply:
This related to this error: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85228/testReport/org.apache.spark.storage/BlockManagerSuite/LRU_with_mixed_storage_levels_and_streams__encryption___on_/
I'm not very sure the reason, but I guess this happens such as follow:
```scala
var i = 0
while (i < inputStream.avaiable()) {
     //do something
}
```
After we arrived at the end of the file which `i == (inputStream.avaiable() - 1)`, then we get `-1` from `inputStream.read()`. And this time we need to call the `refill()` too. Even if we can't get the data from the underlying `fileChannel`, but the `byteBuffer` flipped. So the `inputStream.avaiable` changed, and we still can read the dirty data remained in the `byteBuffer`.

Reply:
Sorry I don't get your point, would you mind share the link to the code block you mention to? Thanks!

Reply:
The problem is when we reach the end of file, we call  `NioBufferedFileInputStream.call()` twice can't get the `-1` both.
you can see here:
org.apache.commons.crypto.stream.input.StreamInput.java
```java
public int read(ByteBuffer dst) throws IOException {
        int remaining = dst.remaining();
        int read = 0;

        while(remaining > 0) {
            int n = this.in.read(this.buf, 0, Math.min(remaining, this.bufferSize));
            if (n == -1) {
                if (read == 0) {
                    read = -1;
                }
                break;
            }

            if (n > 0) {
                dst.put(this.buf, 0, n);
                read += n;
                remaining -= n;
            }
        }

        return read;
    }
```
1. reading the data from the underlying `NioBufferedFileInputStream`, read some remaining bytes.
2. reaching the end, so this time we call the `byteBuffer.flip()`. After `flip()`, the `byteBuffer.hasRemaining()` changed to true, because the `flip()` changed the limit and position.
3. But the return of `StreamInput.read()` is larger than 0 which read from step 1.
4. So we will call the `StreamInput.read()` . But the `byteBuffer.hasRemaining()` of `NioBufferedFileInputStream` changed, so we will read the dirty data from the `byteBuffer`.

I'm not very sure this explanation is correct enough.